File: secure-data-manager/secure-data-manager-backend/src/main/java/ch/post/it/evoting/securedatamanager/setup/protocol/configuration/setupvoting/DeriveCredentialIdAlgorithm.java
/*
 * (c) Copyright 2025 Swiss Post Ltd.
 */
package ch.post.it.evoting.securedatamanager.setup.protocol.configuration.setupvoting;

import static ch.post.it.evoting.cryptoprimitives.utils.ByteArrays.cutToBitLength;
import static ch.post.it.evoting.cryptoprimitives.utils.Conversions.stringToByteArray;
import static ch.post.it.evoting.evotinglibraries.domain.validations.StartVotingKeyValidation.validate;
import static ch.post.it.evoting.evotinglibraries.domain.validations.Validations.validateUUID;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

import ch.post.it.evoting.cryptoprimitives.collection.ImmutableByteArray;
import ch.post.it.evoting.cryptoprimitives.hashing.Argon2;
import ch.post.it.evoting.cryptoprimitives.hashing.Hash;
import ch.post.it.evoting.cryptoprimitives.hashing.HashableList;
import ch.post.it.evoting.cryptoprimitives.hashing.HashableString;
import ch.post.it.evoting.cryptoprimitives.math.Base16;
import ch.post.it.evoting.evotinglibraries.domain.validations.FailedValidationException;

/**
 * Implements the DeriveCredentialId algorithm.
 */
@Service
public class DeriveCredentialIdAlgorithm {

	private final Hash hash;
	private final Base16 base16;
	private final Argon2 argon2;

	public DeriveCredentialIdAlgorithm(final Hash hash,
			final Base16 base16,
			@Qualifier("argon2LessMemory")
			final Argon2 argon2) {
		this.hash = hash;
		this.base16 = base16;
		this.argon2 = argon2;
	}

	/**
	 * Derives a voter's identifier credentialID<sub>id</sub> from the Start Voting Key SVK<sub>id</sub>.
	 *
	 * @param electionEventId ee, the election event id. Must be non-null and a valid UUID.
	 * @param startVotingKey  SVK<sub>id</sub>, the Start Voting Key. Must be non-null and a valid Base32 string of size l<sub>SVK</sub>.
	 * @return the derived credentialID<sub>id</sub>.
	 * @throws NullPointerException      if any parameter is null.
	 * @throws IllegalArgumentException  if the Start Voting Key's size is not
	 *                                   l<sub>SVK</sub>={@value ch.post.it.evoting.evotinglibraries.domain.common.Constants#SVK_LENGTH}.
	 * @throws FailedValidationException if the election event id is not a valid UUID or the start voting key is not a valid Base32 string.
	 */
	@SuppressWarnings("java:S117")
	public String deriveCredentialId(final String electionEventId, final String startVotingKey) {

		// Context.
		final String ee = validateUUID(electionEventId);

		// Input.
		final String SVK_id = validate(startVotingKey);

		// Operation.
		final ImmutableByteArray recursiveHash = hash.recursiveHash(HashableList.of(HashableString.from(ee), HashableString.from("credentialId")));
		final ImmutableByteArray salt = cutToBitLength(recursiveHash, 128);

		final ImmutableByteArray bCredentialID_id = argon2.getArgon2id(stringToByteArray(SVK_id), salt);

		return base16.base16Encode(cutToBitLength(bCredentialID_id, 128));
	}

}

File: voting-client/src/protocol/preliminaries/voter-authentication/derive-credential-id.algorithm.ts
/*
 * (c) Copyright 2025 Swiss Post Ltd.
 */

import {HashService} from "crypto-primitives-ts/lib/esm/hashing/hash_service";
import {Base16Service} from "crypto-primitives-ts/lib/esm/math/base16_service";
import {Argon2Profile} from "crypto-primitives-ts/lib/esm/hashing/argon2_profile";
import {Argon2Service} from "crypto-primitives-ts/lib/esm/hashing/argon2_service";
import {cutToBitLength} from "crypto-primitives-ts/lib/esm/arrays";
import {stringToByteArray} from "crypto-primitives-ts/lib/esm/conversions";
import {ImmutableUint8Array} from "crypto-primitives-ts/lib/esm/immutable_uint8Array";
import {validateUUID} from "../../../domain/validations/validations";
import {validateSVK} from "../../../domain/validations/start-voting-key-validation";
import {DeriveCredentialIdContext} from "./derive-credential-id.types";
import {checkNotNull} from "crypto-primitives-ts/lib/esm/validation/preconditions";

/**
 * Derives a voter's identifier credentialId from the Start Voting Key SVK_id.
 *
 * @param {DeriveCredentialIdContext} context - the DeriveCredentialId context.
 * @param {string} startVotingKey - SVK_id, the start voting key.
 *
 * @return {Promise<string>} - credentialID_id, the derived credentialID_id.
 */
export async function deriveCredentialId(
	context: DeriveCredentialIdContext,
	startVotingKey: string
): Promise<string> {
	// Context.
	checkNotNull(context);
	const ee: string = validateUUID(context.electionEventId);

	// Input.
	const SVK_id: string = validateSVK(startVotingKey);

	const hashService: HashService = new HashService();
	const base16Service: Base16Service = new Base16Service();
	const argon2Service: Argon2Service = new Argon2Service(Argon2Profile.LESS_MEMORY);

	// Operation.
	const salt: ImmutableUint8Array = cutToBitLength(hashService.recursiveHash(ee, "credentialId"), 128);

	const bCredentialID_id: ImmutableUint8Array = await argon2Service.getArgon2id(stringToByteArray(SVK_id), salt);

	// Output.
	// credentialID_id
	return base16Service.base16Encode(cutToBitLength(bCredentialID_id, 128).value());
}
